cohortextractor/dsl_example.py

```diff
@@ -0,0 +1,51 @@
+from cohortextractor.dsl import (
+    Codelist,
+    Cohort,
+    Column,
+    EventTable,
+    PatientTable,
+    categorise,
+)
+
+
+# These table definitions will be imported from elsewhere, and researchers will not have
+# to write them.
+
+
+class Patients(PatientTable):
+    patient_id = Column("patient_id")
+    sex = Column("sex")
+
+
+class Immunisations(EventTable):
+    patient_id = Column("patient_id")
+    date = Column("date")
+    code = Column("code")
+
+
+patients = Patients()
+imms = Immunisations()
+
+# This is a cohort definition, of the kind that researchers will have to implement.
+
+cohort = Cohort()
+cohort.sex = patients.select_column(patients.sex)
+
+covid_imms = imms.filter(imms.code in Codelist(["123", "234"]))
+
+first = covid_imms.sort_by(imms.date).first()
+cohort.first_date = first.select_column(imms.date)
+cohort.first_code = first.select_column(imms.code)
+
+second = (
+    covid_imms.filter(imms.date > (cohort.first_date + 28)).sort_by(imms.date).first()
+)
+cohort.second_date = second.select_column("date")
+cohort.second_code = second.select_column("code")
+
+cohort.codes_match = categorise(
+    {
+        "yes": cohort.first_code == cohort.second_code,
+        "no": True,
+    }
+)
```

Comment:
thinking out loud - ordering is important in this hash, right? i.e. maps to a switch like:
```python
if cohort.first_code == cohort.second_code
  "yes"
elif True
  "no"
```
in which case would it be clearer to use an explicitly ordered structure? ottomh
```python
cohort.codes_match = categorise(
    [  
        ("yes", cohort.first_code == cohort.second_code),
        ("no", True),
    ]
)
```

Reply:
Dicts are now ordered in Python (although relying on that still gives me the willies) so the version in the example is safe.  But it might be better to be more explicit about this and use a list as you propose.

Reply:
thinking of things like this (which looks like a switch) https://github.com/opensafely/nhs-covid-vaccination-coverage/blob/b8111f531960d9ce8c36926f945eb54bdad9dda9/lib/data_processing.py#L138-L152
which would translate to the quite un-aesthetic
```python
cohort.codes_match = categorise(
    {
        "3": (df["care_home"]==1) & (df["age"]>=65),
        "1": (df["age"]>=80),
        "2": (df["age"]>=70),
        "4": (df["shielded"]==1),
        "5": (df["age"]>=65),
        "6": (df["LD"]==1),
        "7": (df["age"]>=60),
        "8": (df["age"]>=55),
        "9": (df["age"]>=50),
        "10": (df["age"]>=40),
        "11": (df["age"]>=30),
        "12": (df["age"]>=18)],
    }
)
```

Reply:
alternatively, we could be encouraging researchers to write explicitly exclusive categories? (sorry if I'm heading off on a tangent here...)

Reply:
I think the list of tuples version is a bit brutal syntactically for people who don't write code all the time.

Reply:
I agree that that's quite un-aesthetic, but it looks a bit better if we drop the `df["..."]`s, which are unneeded here:

```
cohort.codes_match = categorise(
    {
        "3": (care_home == 1) & (age >= 65),
        "1": (age >= 80),
        "2": (age >= 70),
        "4": (shielded == 1),
        "5": (age >= 65),
        "6": (LD == 1),
        "7": (age >= 60),
        "8": (age >= 55),
        "9": (age >= 50),
        "10": (age >= 40),
        "11": (age >= 30),
        "12": (age >= 18),
    }
)
```

Rewritten as a list of `condition`, `value` pairs:

```
cohort.codes_match = categorise(
    [(care_home == 1) & (age >= 65), 3],
    [(age >= 80), 1],
    [(age >= 70), 2],
    [(shielded == 1), 4],
    [(age >= 65), 5],
    [(LD == 1), 6],
    [(age >= 60), 7],
    [(age >= 55), 8],
    [(age >= 50), 9],
    [(age >= 40), 10],
    [(age >= 30), 11],
    [(age >= 18), 12],
)
```

I find that harder to read.

It's easier to read with spacing, but ugly.

```
cohort.codes_match = categorise(
    [(care_home == 1) & (age >= 65),  3],
    [(age >= 80),                     1],
    [(age >= 70),                     2],
    [(shielded == 1),                 4],
    [(age >= 65),                     5],
    [(LD == 1),                       6],
    [(age >= 60),                     7],
    [(age >= 55),                     8],
    [(age >= 50),                     9],
    [(age >= 40),                    10],
    [(age >= 30),                    11],
    [(age >= 18),                    12],
)
```

I'm not sure what the right answer is here, but for now, I don't think it matters too much.